File: Sum Root to Leaf Numbers/Sum_Root_to_Leaf_Numbers.py
# 76ms 15.38%

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        treeStack = []
        lastNode = None
        sum = temp = 0

        while root or len(treeStack) > 0:
            while root:
                temp *= 10
                temp += root.val

                if root.left is None and root.right is None:
                    sum += temp
                treeStack.append(root)

                root = root.left
            if len(treeStack) > 0:
                root = treeStack.pop()

                if root.right is None or root.right == lastNode:
                    temp -= root.val
                    temp /= 10
                    lastNode = root
                    root = None
                else:
                    treeStack.append(root)
                    root = root.right
        return int(sum)

File: Sum Root to Leaf Numbers/Sum_Root_to_Leaf_Numbers.cpp
// 0ms 100%
// 使用后序遍历，并利用一个变量去追踪sum

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include<stack>
using namespace std;

class Solution {
public:
    int sumNumbers(TreeNode* root) {
        stack<TreeNode*> treeStack;
        TreeNode *lastNode = 0;
        int sum = 0, temp = 0;

        while (root || !treeStack.empty())
        {
            while (root)
            {
                temp *= 10;
                temp += root->val;

                if (!root->left && !root->right)
                    sum += temp;   
                treeStack.push(root);
                
                root = root->left;
            }
            if (!treeStack.empty())
            {
                root = treeStack.top();
                treeStack.pop();
                
                if (!root->right || root->right == lastNode)
                {
                    temp -= root->val;
                    temp /= 10;
                    lastNode = root;
                    root = 0;
                }
                else
                {
                    treeStack.push(root);
                    root = root->right;    
                }
            }
        }
        return sum;
    }
};